### PR TITLE
Revert "Revert "Handle tax for split requests"" and fix reported bugs

### PR DESCRIPTION
**Pull request by MonilBhavsar**
Reverts Expensify/App#42670

### Details

Pass taxCode and taxAmount to Split API commands.
Tracks tax both online and offline optimistically for split expense with policy expense chat
### Fixed Issues

$ https://github.com/Expensify/App/issues/39690
PROPOSAL:

### Tests

Same as QA steps

And test https://github.com/Expensify/App/issues/42642

1. Create a split scan.
2. When the receipt is scanning, click on the split preview.
3. Click Amount.
4. Change the currency to foreign currency.
5. Enter amount and save it.
6. Click Tax rate.
7. Ensure no duplicate tax rates are rendered 

And test https://github.com/Expensify/App/issues/42649

1. Create a split scan expense
2. While the receipt is scanning, click on the split preview
3. Click Amount and enter amount
4. Click Tax rate and select a non-zero tax rate
5. Click Tax amount
6. Click Save without modifying the current value
7. Ensure tax amount is saved and no error is displayed

And test https://github.com/Expensify/App/issues/42682
1. Navigate to a workspace chat
2. Click '+' icon > Split expense > Enter amount > Select a tax rate
3. Change the previously selected tax rate with a new rate
4. Verify Tax amount should be updated with the new tax rate

And test https://github.com/Expensify/App/issues/42655

1. Submit expense to any user.
2. Go to transaction thread.
3. Click Amount.
4. Click on the currency.
5. Select a different currency.
6. Verify new currency was saved

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA steps

### QA Steps

Prerequisite: Tax is enabled on workspace

1. Go to FAB > Split > Select a workspace chat
2. Ensure tax rate and tax amount are selected
3. Update currency, expense amount and tax rate and ensure tax rate and tax amount are updated accordingly. Note the taxAmount at this point
4. Split the expense
5. Go to money request with workspace and ensure tax rate is selected as it is and tax amount is divided by half

Now we repeat by splitting with individuals

6. Go to FAB > Split > Select a workspace chat
7. Ensure tax rate and tax amount are not displayed
8. Go to individual money requests and ensure tax rate and tax amount are no displayed
9. Repeat for offline

Testing split scans

10. Go to FAB and select Split Expense and select Scan
11. Add a receipt
12. While in scanning state, update tax rate and ensure tax rate and tax amount are updated in draft transaction accordingly and no API call is being made
13. While in scanning state, update tax amount and ensure tax amount are updated in draft transaction accordingly and no API call is being made
14. While in scanning state, update expense amount and currency and ensure tax rate and tax amount are updated in draft transaction accordingly and no API call is being made

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
See MacOS: Chrome / Safari

</details>

<details>
<summary>Android: mWeb Chrome</summary>
See MacOS: Chrome / Safari

</details>

<details>
<summary>iOS: Native</summary>
See MacOS: Chrome / Safari

</details>

<details>
<summary>iOS: mWeb Safari</summary>
See MacOS: Chrome / Safari

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://drive.google.com/file/d/1C6T9vQ1VwujJJKlCbNNo8cZ_udQZmc3W/view?usp=sharing

Screenshot 2024-05-29 at 4 37 03 PM

Screenshot 2024-05-29 at 4 47 55 PM

</details>

<details>
<summary>MacOS: Desktop</summary>
See MacOS: Chrome / Safari

</details>

### CODE REVIEW

**Comment by MonilBhavsar** [comment]
@dukenv0307 I'm back with a followup PR. I have updated tests section to test and address deploy blockers created on former PR. Please take a look, thank you!

**Comment by MonilBhavsar** [comment]
@dukenv0307 sorry, I am not understanding it. Could you please explain it again. In Video, i see you're only updating taxAmount?

**Comment by MonilBhavsar** [comment]
This happens on staging too and is specific to currencies like VND where we don't support decimals. I would address that in another issue and PR

**Comment by trjExpensify** [comment]
With that cleared up, @dukenv0307. Can you complete the review?

File: src/components/MoneyRequestConfirmationList.tsx [hunk 1]
```diff
@@ -345,6 +339,31 @@ function MoneyRequestConfirmationList({
...
+    useEffect(() => {
+        if (!shouldShowTax) {
+            return;
+        }
+        setTaxAmountInDraft(transaction, policy);
+        // eslint-disable-next-line react-hooks/exhaustive-deps -- we want to call this function when component is mounted
+    }, []);
+
+    useEffect(() => {
+        if (!shouldShowTax || prevTaxAmount !== transaction?.taxAmount || (prevAmount === transaction?.amount && prevCurrency === transaction?.currency)) {
```
[the remark this answers is not in the document]
Comment (MonilBhavsar) [reply src/components/MoneyRequestConfirmationList.tsx:351]: What do you think should be expected behavior here?

[the remark this answers is not in the document]
Comment (MonilBhavsar) [reply src/components/MoneyRequestConfirmationList.tsx:351]: Ah okay, I got you're saying now. Looking

[the remark this answers is not in the document]
Comment (MonilBhavsar) [reply src/components/MoneyRequestConfirmationList.tsx:351]: @dukenv0307 I have fixed this bug. Hopefully, I didn't break anything in fixing it 🙏

**Review by dukenv0307 — APPROVED** [review]

**Comment by MonilBhavsar** [comment]
Resolved conflicts

**Comment by dangrous** [comment]
I can look today but now there are conflicts - I'll start my review and see how far I get, and can finish up post-conflicts tomorrow!

**Comment by MonilBhavsar** [comment]
Okay thanks! I'll make sure to resolve it before Mountain time, so we don't get it again 😅

**Review by dangrous — DISMISSED** [review]
Okay this looks good to me! I can take another final look (and merge) once conflicts are resolved.

**Comment by MonilBhavsar** [comment]
@dangrous thanks! Conflicts resolved. Doing a quick test

**Comment by MonilBhavsar** [comment]
tested again and works fine!
In last commit, I moved code around, so it does not show up in difference unnecessarily

**Review by dangrous — APPROVED** [review]
Looks good! Do we think @dukenv0307 should give it one more set of tests, since there have been some changes since last time, or am I good to merge?

**Comment by MonilBhavsar** [comment]
Okay cool! @dukenv0307 could you please take a final look, please 🙇

**Comment by trjExpensify** [comment]
Were the changes material?

**Comment by MonilBhavsar** [comment]
Nice, thanks for checking!

File: src/pages/iou/request/step/IOURequestStepAmount.tsx [hunk 2]
```diff
@@ -273,25 +278,25 @@ function IOURequestStepAmount({
         }
 
         // If the value hasn't changed, don't request to save changes on the server and just close the modal
-        const transactionCurrency = TransactionUtils.getCurrency(transaction);
-        if (newAmount === TransactionUtils.getAmount(transaction) && currency === transactionCurrency) {
+        const transactionCurrency = TransactionUtils.getCurrency(currentTransaction);
+        if (newAmount === TransactionUtils.getAmount(currentTransaction) && currency === transactionCurrency) {
```
Comment (rlinoz) [inline src/pages/iou/request/step/IOURequestStepAmount.tsx:282]: Changing how to get the transaction caused this blocker https://github.com/Expensify/App/issues/43259

**Pull request merged**